modules/effects/src/main/scala/dev/profunktor/redis4cats/transactions.scala

```diff
@@ -24,16 +24,29 @@ import dev.profunktor.redis4cats.effect.Log
 
 object transactions {
 
-  case class RedisTransaction[F[_]: Log: Sync, K, V](
+  case class RedisTransaction[F[_]: Concurrent: Log, K, V](
       cmd: RedisCommands[F, K, V]
   ) {
-    def run[A](fa: F[A]): F[A] =
-      Log[F].info("Transaction started") *>
-        cmd.multi.bracketCase(_ => fa) {
+
+    def run(fas: F[Any]*): F[Unit] = {
+      val tx =
+        Resource.makeCase(cmd.multi) {
           case (_, ExitCase.Completed) => cmd.exec *> Log[F].info("Transaction completed")
           case (_, ExitCase.Error(e))  => cmd.discard *> Log[F].error(s"Transaction failed: ${e.getMessage}")
           case (_, ExitCase.Canceled)  => cmd.discard *> Log[F].error("Transaction canceled")
         }
+
+      val commands =
+        Resource.makeCase(fas.toList.traverse(_.start)) {
+          case (_, ExitCase.Completed) => ().pure[F]
+          case (fb, ExitCase.Error(_)) => fb.traverse_(_.cancel)
+          case (fb, ExitCase.Canceled) => fb.traverse_(_.cancel)
+        }
+
+      Log[F].info("Transaction started") *>
+        commands.flatTap(_ => tx).use(_.traverse_(_.join))
```

Comment:
If I understand correctly:

Happy path:

1. Log "Transaction started"
2. `commands` is allocated - all `fas` are started in parallel
3. `tx` is allocated - `cmd.multi` is executed
4. `_.traverse(_.join)` - waiting for commands to complete
5. `tx` is cleaned up - `cmd.exec` and `Transaction completed` logged
6. `commands` is cleaned up - no-op

It feels weird that `cmd.multi` is executed after `fas` are started...

Reply:
@gvolpe can you give me a quick explanation on how transactions work under the hood? Maybe I'll be able to come up with something cleaner

Reply:
AFAIU `multi` should run first, then all the other operations, then `exec` / `discard` (based on https://redis.io/commands/multi).

In that case I'd remove `commands`, do `tx.use(_ => fas.parSequence_)` and call it a day...

* - `parSequence` might be replaced with a `.background`-based implementation if we need to use Concurrent instead of Parallel - like `fas.traverse(_.background).use(_.sequence_)`

Reply:
@kubukoz you are right, `MULTI` should run first. Though, that `parSequence` doesn't seem to work with the current API, I'll need to play with this a bit more but it'll have to wait until Monday.